core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -287,15 +289,65 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
     def appendCallback(error: Errors): Unit =
       error match {
         case Errors.NONE =>
-          trace(s"Completed transaction for ${txnLogAppend.transactionalId} with coordinator epoch ${txnLogAppend.coordinatorEpoch}, final state: state after commit: ${txnLogAppend.txnMetadata.state}")
+          trace(s"Completed transaction for ${txnLogAppend.transactionalId} with coordinator epoch ${txnLogAppend.coordinatorEpoch}, final state after commit: ${txnLogAppend.txnMetadata.state}")
 
         case Errors.NOT_COORDINATOR =>
           info(s"No longer the coordinator for transactionalId: ${txnLogAppend.transactionalId} while trying to append to transaction log, skip writing to transaction log")
 
         case Errors.COORDINATOR_NOT_AVAILABLE =>
-          warn(s"Failed updating transaction state for ${txnLogAppend.transactionalId} when appending to transaction log due to ${error.exceptionName}. retrying")
-          // enqueue for retry
-          txnLogAppendRetryQueue.add(txnLogAppend)
+          info(s"I am currently not available to append $txnLogAppend: possible causes include ${Errors.UNKNOWN_TOPIC_OR_PARTITION}, ${Errors.NOT_ENOUGH_REPLICAS}, ${Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND} and ${Errors.REQUEST_TIMED_OUT}")
+
+          val transactionalId = txnLogAppend.transactionalId
+          val coordinatorEpoch = txnLogAppend.coordinatorEpoch
+
+          txnStateManager.getTransactionState(transactionalId) match {
```

Comment:
I had thought about this (see my other comment). The detailed reasoning is that:

1. For `NOT_COORDINATOR`, we would abort appending and reset the pending state just in case (the metadata should be removed from the cache anyways);

2. For `COORDINATOR_NOT_AVAILBLE`, we may also want to check the metadata is still validate, and if not also abort the appending again. If we retry again it may then be changed to `NOT_COORDINATOR` but we may want to fail-fast.